src/server/ua_server_binary.c

```diff
@@ -283,84 +289,55 @@ processOPN(UA_Server *server, UA_Connection *connection,
-    UA_SequenceHeader seqHeader;
     UA_NodeId requestType;
-    UA_OpenSecureChannelRequest r;
+    UA_OpenSecureChannelRequest openSecureChannelRequest;
     size_t offset = 0;
-    retval |= UA_AsymmetricAlgorithmSecurityHeader_decodeBinary(msg, &offset, &asymHeader);
-    retval |= UA_SequenceHeader_decodeBinary(msg, &offset, &seqHeader);
     retval |= UA_NodeId_decodeBinary(msg, &offset, &requestType);
-    retval |= UA_OpenSecureChannelRequest_decodeBinary(msg, &offset, &r);
+    retval |= UA_OpenSecureChannelRequest_decodeBinary(msg, &offset, &openSecureChannelRequest);
 
     /* Error occured */
-    if(retval != UA_STATUSCODE_GOOD || requestType.identifier.numeric != 446) {
-        UA_AsymmetricAlgorithmSecurityHeader_deleteMembers(&asymHeader);
+    if(retval != UA_STATUSCODE_GOOD || requestType.identifier.numeric != UA_TYPES[UA_TYPES_OPENSECURECHANNELREQUEST].binaryEncodingId) {
         UA_NodeId_deleteMembers(&requestType);
-        UA_OpenSecureChannelRequest_deleteMembers(&r);
+        UA_OpenSecureChannelRequest_deleteMembers(&openSecureChannelRequest);
         connection->close(connection);
         return;
     }
 
     /* Call the service */
-    UA_OpenSecureChannelResponse p;
-    UA_OpenSecureChannelResponse_init(&p);
-    Service_OpenSecureChannel(server, connection, &r, &p);
-    UA_OpenSecureChannelRequest_deleteMembers(&r);
+    UA_OpenSecureChannelResponse openScResponse;
+    UA_OpenSecureChannelResponse_init(&openScResponse);
+    Service_OpenSecureChannel(server, connection, &openSecureChannelRequest, &openScResponse, preparedChannel);
+    UA_OpenSecureChannelRequest_deleteMembers(&openSecureChannelRequest);
 
-    /* Opening the channel failed */
+    // Opening the channel failed
     UA_SecureChannel *channel = connection->channel;
     if(!channel) {
-        UA_OpenSecureChannelResponse_deleteMembers(&p);
-        UA_AsymmetricAlgorithmSecurityHeader_deleteMembers(&asymHeader);
+        UA_OpenSecureChannelResponse_deleteMembers(&openScResponse);
         connection->close(connection);
         return;
     }
 
-    /* Set the starting sequence number */
-    channel->receiveSequenceNumber = seqHeader.sequenceNumber;
+    UA_ByteString_copy(&server->config.serverCertificate, &channel->localAsymAlgSettings.senderCertificate);
+    UA_ByteString_copy(&channel->securityPolicy->policyUri, &channel->localAsymAlgSettings.securityPolicyUri);
 
-    /* Allocate the return message */
-    UA_ByteString resp_msg;
-    UA_ByteString_init(&resp_msg);
-    retval = connection->getSendBuffer(connection, connection->localConf.sendBufferSize, &resp_msg);
-    if(retval != UA_STATUSCODE_GOOD) {
-        UA_OpenSecureChannelResponse_deleteMembers(&p);
-        UA_AsymmetricAlgorithmSecurityHeader_deleteMembers(&asymHeader);
-        connection->close(connection);
-        return;
-    }
+    UA_ByteString_allocBuffer(&channel->localAsymAlgSettings.receiverCertificateThumbprint,
```

Comment:
Why allocBuffer instead of getSendBuffer?
- There is a rationale behind the networklayer allocating memory. This can point directly into the network card.
- allocBuffer may fail as well. Check all return values and clean up.